### DISCUSSION TITLE
Component for onboarding tour popovers?

### DISCUSSION THREAD

**Discussion by sampl** [radix-ui/primitives#1199] [Ideas] [answered]
Hey all,

Is it currently possible to combine a modal popover with a dialog-style overlay?

This is a common pattern in new user onboarding tours--admittedly an annoying one sometimes, but I think there are legit use cases.

I could probably hack it by adding a massive box shadow to a modal popover (`box-shadow: 0 0 0 9999px hsla(0 0% 0% .5)`) but ideally the solution would let the target appear "above" the overlay as in the examples, and even stay interactive.

Existing standalone react libs that do this:
- https://react-joyride.com/
- https://reactour.vercel.app/

Maybe this would be an optional overlay component on popover? Or if it was a new component the API might look like this (just spitballin'):

```jsx
import * as TourPoint from '@radix-ui/react-tour-point';

export default () => (
  <TourPoint.Root>
    <TourPoint.Trigger />
    <TourPoint.Portal>
      <TourPoint.Overlay />
      <TourPoint.Content>
        <TourPoint.Title />
        <TourPoint.Description />
        <TourPoint.Close />
        <TourPoint.Arrow />
      </TourPoint.Content>
    </TourPoint.Portal>
  </TourPoint.Root>
);
```
 
Real world examples:

<img width="985" alt="Screen Shot 2022-02-24 at 3 21 20 PM" src="https://user-images.githubusercontent.com/231851/155601330-bebe4985-4f56-4591-92b9-8857e8556b22.png">

![5ffdc0b94914fbd8e36ed732_linkedin-tooltips](https://user-images.githubusercontent.com/231851/155601186-5ddd59bb-6a50-425a-9f2d-0ebc2a4e4466.jpg)

**Comment by benoitgrelard** [comment]
Hey @sampl this would definitely be cool to build. It seems the solutions you link to have a ton of cool features.

You could probably hack something around right now using `Dialog` and `Popover`, but I'm not too sure if it would be very accessible…

It seems this pattern alltogether would need quite a bit of research to see how it can be made accessible seeing that it needs to "isolate" portions of the rendered page rather than separate modal content.

But I am sure it would be possible.

Thanks for the suggestion!

✌️

**Reply by sampl** [reply]
Makes sense, thanks for your take @benoitgrelard :)

**Comment by dcastil** [comment]
I also have a need for an overlay for the popover, select, etc. from Radix. I work on an app where the CSS `pointer-events: auto` is used in some elements, so the `modal` property doesn't work as expected and doesn't disable those elements. Rendering a transparent overlay on top of the UI would also make those elements non-clickable.

**Comment by sampl** [comment, marked as answer]
FWIW I ended up doing this:

- Use a radix popover with `modal="true"`
- Add a raw `div` for the backdrop (not a Radix `Dialog.Overlay`, just a styled div) and only render it if the popover is open
- For the thing I want to appear "above" the backdrop: give it a `z-index: 99999;` but only when the popover is open

I can make a codepen later if yall want it

EDIT: Codepen here: https://codesandbox.io/s/festive-grass-21wlxo

**Reply by dcastil** [reply]
I also thought about something like that. I guess you made the component controlled for that reason, right?

**Reply by marcelobaez** [reply]
I'm late to the party. Would you mind share a codepen of your implementation?

**Reply by sampl** [reply]
Yup, it's controlled. Sorry for the delay @marcelobaez here's a codesandbox: https://codesandbox.io/s/festive-grass-21wlxo

**Reply by Code-Victor** [reply]
@sampl great solution.
How were you able to make it work for many elements?

**Reply by sampl** [reply]
@Code-Victor can you be more specific? Check the codesandbox link and explanation above. Basically, for any element you want to appear visible, you temporarily give it a z-index higher than the overlay z-index--could be as many as you want.

**Reply by Code-Victor** [reply]
@sampl  Thanks, what I wanted to know is
1. How can you make this work for various elements of your onboarding process?
2. Were you able to wrap each element you wanted to onboard individually?

**Reply by sampl** [reply]
@Code-Victor yes you can use this for multiple elements. It's controlled with React state, so you'll have to make it work with whatever your app's existing onboarding state is. Consider what flags and user data conditions you want to show the tour in. Sorry, that's as much as I can help.

**Reply by Code-Victor** [reply]
Thank you, I think I can work with that.

**Reply by itiurin-doto** [reply]
can you please share the link to codesandbox ? @sampl

**Reply by sampl** [reply]
@itiurin-doto https://codesandbox.io/s/festive-grass-21wlxo
